presto-main/src/main/java/com/facebook/presto/operator/scalar/MapToMapCast.java

```diff
@@ -13,70 +13,181 @@
+
+    private static final MethodHandle METHOD_HANDLE = methodHandle(
+            MapToMapCast.class,
+            "mapCast",
+            MethodHandle.class,
+            MethodHandle.class,
+            Type.class,
+            ConnectorSession.class,
+            Block.class);
+
+    public MapToMapCast()
+    {
+        super(CAST,
+                ImmutableList.of(typeVariable("FK"), typeVariable("FV"), typeVariable("TK"), typeVariable("TV")),
+                ImmutableList.of(),
+                parseTypeSignature("map(TK,TV)"),
+                ImmutableList.of(parseTypeSignature("map(FK,FV)")));
+    }
+
+    @Override
+    public ScalarFunctionImplementation specialize(BoundVariables boundVariables, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
+    {
+        checkArgument(arity == 1, "Expected arity to be 1");
+        Type fromKeyType = boundVariables.getTypeVariable("FK");
+        Type fromValueType = boundVariables.getTypeVariable("FV");
+        Type toKeyType = boundVariables.getTypeVariable("TK");
+        Type toValueType = boundVariables.getTypeVariable("TV");
+        Type toMapType = MAP.createType(typeManager, ImmutableList.of(TypeParameter.of(toKeyType), TypeParameter.of(toValueType)));
+
+        MethodHandle keyProcessor = getProcessMethodHandle(functionRegistry, fromKeyType, toKeyType, true);
+        MethodHandle valueProcessor = getProcessMethodHandle(functionRegistry, fromValueType, toValueType, false);
+        MethodHandle target = MethodHandles.insertArguments(METHOD_HANDLE, 0, keyProcessor, valueProcessor, toMapType);
+        return new ScalarFunctionImplementation(true, ImmutableList.of(valueTypeArgumentProperty(RETURN_NULL_ON_NULL)), target, true);
+    }
+
+    private MethodHandle getProcessMethodHandle(FunctionRegistry functionRegistry, Type fromType, Type toType, boolean isKey)
+    {
+        MethodHandle getter = nativeValueGetter(fromType);
+
+        // Adapt cast that takes ([ConnectorSession,] ?) to one that takes (?, ConnectorSession), where ? is the return type of getter.
+        MethodHandle cast = functionRegistry.getScalarFunctionImplementation(internalOperator(CAST, toType, ImmutableList.of(fromType))).getMethodHandle();
+        if (cast.type().parameterArray()[0] != ConnectorSession.class) {
+            cast = MethodHandles.dropArguments(cast, 0, ConnectorSession.class);
+        }
+        cast = permuteArguments(cast, methodType(cast.type().returnType(), cast.type().parameterArray()[1], cast.type().parameterArray()[0]), 1, 0);
+        MethodHandle target = compose(cast, getter);
+
+        // We only need to check null if the method handle is for key and the cast result is nullable.
+        Class<?> castResultJavaType = target.type().returnType();
+        if (isKey && !castResultJavaType.isPrimitive()) {
+            MethodHandle nullCheck;
+            if (castResultJavaType == Long.class) {
```

Comment:
Put this `if` in a separate method. Probably a private `nullChecker` method in this class. Add javadoc to that method to explain what it does, and what the signature is.